File: src/chat/chat.service.ts
import { CACHE_MANAGER } from '@nestjs/cache-manager';
import {
  BadRequestException,
  ForbiddenException,
  Inject,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import * as cacheManager from 'cache-manager';

import { PrismaService } from '@/prisma/prisma.service';

@Injectable()
export class ChatService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly configService: ConfigService,
    @Inject(CACHE_MANAGER) private cacheManager: cacheManager.Cache,
  ) {}

  async getSessionData(sessionId: string) {
    const sessionDataStr = await this.cacheManager.get<string>(
      `session:${sessionId}`,
    );
    if (!sessionDataStr) {
      return null;
    }
    return JSON.parse(sessionDataStr);
  }

  async getUserById(userId: number) {
    return await this.prisma.user.findUnique({
      where: { id: userId },
      include: { role: true },
    });
  }

  // Начать чат по товару
  async startChat(productId: number, buyerId: number) {
    // Проверяем, что товар существует
    const product = await this.prisma.product.findUnique({
      where: { id: productId },
      include: {
        user: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
      },
    });

    if (!product) {
      throw new NotFoundException('Товар не найден');
    }

    // Проверяем, что покупатель не пытается написать самому себе
    if (product.userId === buyerId) {
      throw new BadRequestException('Нельзя писать самому себе');
    }

    // Проверяем, существует ли уже чат между этими пользователями по этому товару
    const existingChat = await this.prisma.chat.findFirst({
      where: {
        productId,
        buyerId,
        sellerId: product.userId,
      },
      include: {
        product: {
          select: {
            id: true,
            name: true,
            images: true,
            price: true,
          },
        },
        seller: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
        buyer: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
      },
    });

    if (existingChat) {
      return {
        ...existingChat,
        product: existingChat.product
          ? {
              ...existingChat.product,
              images: existingChat.product.images.map((image) => {
                image;
              }),
            }
          : null,
      };
    }

    // Создаем новый чат
    const newChat = await this.prisma.chat.create({
      data: {
        productId,
        buyerId,
        sellerId: product.userId,
      },
      include: {
        product: {
          select: {
            id: true,
            name: true,
            images: true,
            price: true,
          },
        },
        seller: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
        buyer: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
      },
    });

    return {
      ...newChat,
      product: newChat.product
        ? {
            ...newChat.product,
            images: newChat.product.images.map((image) => `${image}`),
          }
        : null,
    };
  }

  // Получить все чаты пользователя
  async getUserChats(userId: number) {
    const chats = await this.prisma.chat.findMany({
      where: {
        OR: [{ buyerId: userId }, { sellerId: userId }],
      },
      include: {
        product: {
          select: {
            id: true,
            name: true,
            images: true,
            price: true,
          },
        },
        buyer: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
        seller: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
        lastMessage: {
          select: {
            id: true,
            content: true,
            createdAt: true,
            senderId: true,
            isRead: true,
          },
        },
      },
      orderBy: {
        lastMessageAt: 'desc',
      },
    });

    // Форматируем данные для фронтенда
    return chats.map((chat) => {
      const isUserBuyer = chat.buyerId === userId;
      const companion = isUserBuyer ? chat.seller : chat.buyer;
      const unreadCount = isUserBuyer
        ? chat.unreadCountBuyer
        : chat.unreadCountSeller;

      // Форматируем дату в dd.mm.yy
      const formatDate = (date: Date) => {
        const day = date.getDate().toString().padStart(2, '0');
        const month = (date.getMonth() + 1).toString().padStart(2, '0');
        const year = date.getFullYear().toString().slice(-2);
        return `${day}.${month}.${year}`;
      };

      return {
        id: chat.id,
        isModerationChat: chat.isModerationChat, // Признак чата модерации
        // Информация о товаре (может быть null для чатов модерации)
        product: chat.product
          ? {
              id: chat.product.id,
              name: chat.product.name,
              price: chat.product.price,
              image: chat.product.images[0]
                ? `${chat.product.images[0]}`
                : null, // Первое фото товара
            }
          : null,
        // Информация о собеседнике (продавце для покупателя, покупателе для продавца)
        companion: {
          id: companion.id,
          fullName: companion.fullName,
          phoneNumber: companion.phoneNumber,
          // TODO: Добавить аватарку когда будет поле в User
          avatar: null,
        },
        // Последнее сообщение
        lastMessage: chat.lastMessage
          ? {
              content: chat.lastMessage.content,
              createdAt: chat.lastMessage.createdAt,
              formattedDate: formatDate(chat.lastMessage.createdAt),
              isFromMe: chat.lastMessage.senderId === userId,
              isRead: chat.lastMessage.isRead,
            }
          : null,
        // Счетчик непрочитанных
        unreadCount,
        // Дата последнего обновления чата
        lastActivity: formatDate(chat.lastMessageAt || chat.createdAt),
        createdAt: chat.createdAt,
      };
    });
  }

  // Получить информацию о чате с деталями товара и продавца
  async getChatInfo(chatId: number, userId: number) {
    const chat = await this.prisma.chat.findUnique({
      where: { id: chatId },
      include: {
        product: {
          select: {
            id: true,
            name: true,
            images: true,
            price: true,
            description: true,
          },
        },
        buyer: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
        seller: {
          select: {
            id: true,
            fullName: true,
            phoneNumber: true,
          },
        },
      },
    });

    if (!chat) {
      throw new NotFoundException('Чат не найден');
    }

    // Проверяем доступ к чату
    if (chat.buyerId !== userId && chat.sellerId !== userId) {
      throw new ForbiddenException('Нет доступа к этому чату');
    }

    const isUserBuyer = chat.buyerId === userId;
    const companion = isUserBuyer ? chat.seller : chat.buyer;

    return {
      id: chat.id,
      isModerationChat: chat.isModerationChat,
      // Информация о товаре (отображается сверху в чате, может быть null для чатов модерации)
      product: chat.product
        ? {
            id: chat.product.id,
            name: chat.product.name,
            price: chat.product.price,
            image: chat.product.images[0] ? `${chat.product.images[0]}` : null,
            description: chat.product.description,
          }
        : null,
      // Информация о собеседнике (продавце)
      companion: {
        id: companion.id,
        fullName: companion.fullName,
        phoneNumber: companion.phoneNumber,
        role: isUserBuyer ? 'seller' : 'buyer',
      },
      // Мета информация
      isUserBuyer,
      unreadCount: isUserBuyer ? chat.unreadCountBuyer : chat.unreadCountSeller,
      createdAt: chat.createdAt,
    };
  }

  // Получить сообщения чата с пагинацией
  async getChatMessages(
    chatId: number,
    userId: number,
    page: number = 1,
    limit: number = 50,
  ) {
    // Проверяем доступ к чату
    const chat = await this.prisma.chat.findUnique({
      where: { id: chatId },
      select: { buyerId: true, sellerId: true },
    });

    if (!chat) {
      throw new NotFoundException('Чат не найден');
    }

    if (chat.buyerId !== userId && chat.sellerId !== userId) {
      throw new ForbiddenException('Нет доступа к этому чату');
    }

    // Получаем сообщения с пагинацией
    const messages = await this.prisma.message.findMany({
      where: { chatId },
      include: {
        sender: {
          select: {
            id: true,
            fullName: true,
          },
        },
        relatedProduct: {
          select: {
            id: true,
            name: true,
            images: true,
            price: true,
            moderateState: true,
          },
        },
      },
      orderBy: {
        createdAt: 'desc',
      },
      skip: (page - 1) * limit,
      take: limit,
    });

    // Форматируем сообщения
    const formattedMessages = messages.reverse().map((message) => ({
      id: message.id,
      content: message.content,
      senderId: message.senderId,
      sender: message.sender,
      relatedProduct: message.relatedProduct, // Добавляем информацию о связанном товаре
      isFromMe: message.senderId === userId,
      isRead: message.isRead,
      readAt: message.readAt,
      createdAt: message.createdAt,
      // Время отправки в удобном формате
      timeString: message.createdAt.toLocaleTimeString('ru-RU', {
        hour: '2-digit',
        minute: '2-digit',
      }),
    }));

    // Подсчитываем общее количество сообщений
    const total = await this.prisma.message.count({
      where: { chatId },
    });

    return {
      messages: formattedMessages,
      pagination: {
        page,
        limit,
        total,
        pages: Math.ceil(total / limit),
      },
    };
  }

  // Отправить сообщение
  async sendMessage(chatId: number, senderId: number, content: string) {
    // Проверяем доступ к чату
    const chat = await this.prisma.chat.findUnique({
      where: { id: chatId },
    });

    if (!chat) {
      throw new NotFoundException('Чат не найден');
    }

    if (chat.buyerId !== senderId && chat.sellerId !== senderId) {
      throw new ForbiddenException('Нет доступа к этому чату');
    }

    // Создаем сообщение
    const message = await this.prisma.message.create({
      data: {
        content,
        senderId,
        chatId,
      },
      include: {
        sender: {
          select: {
            id: true,
            fullName: true,
          },
        },
      },
    });

    // Обновляем счетчики непрочитанных сообщений и последнее сообщение
    const isFromBuyer = senderId === chat.buyerId;

    await this.prisma.chat.update({
      where: { id: chatId },
      data: {
        lastMessageId: message.id,
        lastMessageAt: new Date(),
        // Увеличиваем счетчик непрочитанных для получателя
        ...(isFromBuyer
          ? { unreadCountSeller: { increment: 1 } }
          : { unreadCountBuyer: { increment: 1 } }),
      },
    });

    return {
      id: message.id,
      content: message.content,
      senderId: message.senderId,
      sender: message.sender,
      isRead: message.isRead,
      createdAt: message.createdAt,
      timeString: message.createdAt.toLocaleTimeString('ru-RU', {
        hour: '2-digit',
        minute: '2-digit',
      }),
    };
  }

  // Отправить сообщение с привязкой к товару (для чатов модерации)
  async sendMessageWithProduct(
    chatId: number,
    senderId: number,
    content: string,
    productId: number,
  ) {
    // Проверяем доступ к чату
    const chat = await this.prisma.chat.findUnique({
      where: { id: chatId },
    });

    if (!chat) {
      throw new NotFoundException('Чат не найден');
    }

    if (chat.buyerId !== senderId && chat.sellerId !== senderId) {
      throw new ForbiddenException('Нет доступа к этому чату');
    }

    // Создаем сообщение с привязкой к товару
    const message = await this.prisma.message.create({
      data: {
        content,
        senderId,
        chatId,
        relatedProductId: productId, // Привязываем к товару
      },
      include: {
        sender: {
          select: {
            id: true,
            fullName: true,
          },
        },
        relatedProduct: {
          select: {
            id: true,
            name: true,
            images: true,
            price: true,
          },
        },
      },
    });

    // Обновляем счетчики непрочитанных сообщений и последнее сообщение
    const isFromBuyer = senderId === chat.buyerId;

    await this.prisma.chat.update({
      where: { id: chatId },
      data: {
        lastMessageId: message.id,
        lastMessageAt: new Date(),
        // Увеличиваем счетчик непрочитанных для получателя
        ...(isFromBuyer
          ? { unreadCountSeller: { increment: 1 } }
          : { unreadCountBuyer: { increment: 1 } }),
      },
    });

    return {
      id: message.id,
      content: message.content,
      senderId: message.senderId,
      sender: message.sender,
      relatedProduct: message.relatedProduct,
      isRead: message.isRead,
      createdAt: message.createdAt,
      timeString: message.createdAt.toLocaleTimeString('ru-RU', {
        hour: '2-digit',
        minute: '2-digit',
      }),
    };
  }

  // Отметить сообщения как прочитанные
  async markMessagesAsRead(chatId: number, userId: number) {
    // Проверяем доступ к чату
    const chat = await this.prisma.chat.findUnique({
      where: { id: chatId },
    });

    if (!chat) {
      throw new NotFoundException('Чат не найден');
    }

    if (chat.buyerId !== userId && chat.sellerId !== userId) {
      throw new ForbiddenException('Нет доступа к этому чату');
    }

    // Отмечаем все непрочитанные сообщения как прочитанные
    await this.prisma.message.updateMany({
      where: {
        chatId,
        senderId: { not: userId }, // Не свои сообщения
        isRead: false,
      },
      data: {
        isRead: true,
        readAt: new Date(),
      },
    });

    // Обнуляем счетчик непрочитанных сообщений
    const isUserBuyer = chat.buyerId === userId;

    await this.prisma.chat.update({
      where: { id: chatId },
      data: isUserBuyer ? { unreadCountBuyer: 0 } : { unreadCountSeller: 0 },
    });

    return { success: true, message: 'Сообщения отмечены как прочитанные' };
  }
}

File: api-go/internal/service/chat_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"math"
	"slices"
	"strings"
	"time"

	"med-vito/api-go/internal/config"
	"med-vito/api-go/internal/pkg/mail"
	"med-vito/api-go/internal/repository"
)

type ChatService struct {
	repo  *repository.ChatPG
	users *repository.UserPG
	cfg   config.Config
}

func NewChatService(repo *repository.ChatPG, users *repository.UserPG, cfg config.Config) *ChatService {
	return &ChatService{repo: repo, users: users, cfg: cfg}
}

func formatChatShortDate(t time.Time) string {
	lt := t.In(time.Local)
	return fmt.Sprintf("%02d.%02d.%02d", lt.Day(), int(lt.Month()), lt.Year()%100)
}

func messageTimeString(t time.Time) string {
	return t.In(time.Local).Format("15:04")
}

func (s *ChatService) formatStartChat(c *repository.ChatFullRow) map[string]any {
	var product any
	if c.JoinProductID != nil {
		product = map[string]any{
			"id":     *c.JoinProductID,
			"name":   c.ProductName,
			"images": append([]string{}, c.ProductImages...),
			"price":  c.ProductPrice,
		}
	} else {
		product = nil
	}
	return map[string]any{
		"id":               c.ID,
		"productId":        c.ProductID,
		"buyerId":          c.BuyerID,
		"sellerId":         c.SellerID,
		"isModerationChat": c.IsModerationChat,
		"lastMessageAt":    c.LastMessageAt,
		"createdAt":        c.CreatedAt,
		"product":          product,
		"buyer":            map[string]any{"id": c.BuyerID, "fullName": c.BuyerName, "phoneNumber": c.BuyerPhone},
		"seller":           map[string]any{"id": c.SellerID, "fullName": c.SellerName, "phoneNumber": c.SellerPhone},
	}
}

func (s *ChatService) StartChat(ctx context.Context, productID, buyerID int32) (map[string]any, error) {
	sellerID, ok, err := s.repo.ProductSellerID(ctx, productID)
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, &AppError{404, "Товар не найден"}
	}
	if sellerID == buyerID {
		return nil, &AppError{400, "Нельзя писать самому себе"}
	}
	existing, err := s.repo.FindChatByProductParticipants(ctx, productID, buyerID, sellerID)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return s.formatStartChat(existing), nil
	}
	newID, err := s.repo.InsertProductChat(ctx, productID, buyerID, sellerID)
	if err != nil {
		return nil, err
	}
	full, err := s.repo.GetChatFull(ctx, newID)
	if err != nil {
		return nil, err
	}
	return s.formatStartChat(full), nil
}

func (s *ChatService) StartDirectChat(ctx context.Context, sellerID, buyerID int32) (map[string]any, error) {
	if sellerID <= 0 {
		return nil, &AppError{400, "Нужен sellerId"}
	}
	if sellerID == buyerID {
		return nil, &AppError{400, "Нельзя писать самому себе"}
	}
	existing, err := s.repo.FindDirectChatBetweenUsers(ctx, buyerID, sellerID)
	if err != nil {
		return nil, err
	}
	if existing != nil {
		return s.formatStartChat(existing), nil
	}
	newID, err := s.repo.InsertDirectChat(ctx, buyerID, sellerID)
	if err != nil {
		return nil, err
	}
	full, err := s.repo.GetChatFull(ctx, newID)
	if err != nil {
		return nil, err
	}
	return s.formatStartChat(full), nil
}

func (s *ChatService) GetUserChats(ctx context.Context, userID int32) ([]map[string]any, error) {
	rows, err := s.repo.ListChatsForUser(ctx, userID)
	if err != nil {
		return nil, err
	}
	out := make([]map[string]any, 0, len(rows))
	for _, chat := range rows {
		isBuyer := chat.BuyerID == userID
		var compName, compPhone string
		var compID int32
		if isBuyer {
			compID, compName, compPhone = chat.SellerID, chat.SellerName, chat.SellerPhone
		} else {
			compID, compName, compPhone = chat.BuyerID, chat.BuyerName, chat.BuyerPhone
		}
		if chat.IsModerationChat {
			compName = "Модерация платформы"
			compPhone = ""
		}
		unread := chat.UnreadSeller
		if isBuyer {
			unread = chat.UnreadBuyer
		}
		var product any
		if chat.JoinProductID != nil {
			var img any
			if len(chat.ProductImages) > 0 {
				img = chat.ProductImages[0]
			} else {
				img = nil
			}
			product = map[string]any{
				"id":    *chat.JoinProductID,
				"name":  chat.ProductName,
				"price": chat.ProductPrice,
				"image": img,
			}
		} else {
			product = map[string]any{
				"id":    compID,
				"name":  "Чат с " + compName,
				"price": 0,
				"image": nil,
			}
		}
		var lastMsg any
		if chat.LMID != nil && chat.LMContent != nil && chat.LMCreated != nil && chat.LMSender != nil && chat.LMRead != nil {
			lastMsg = map[string]any{
				"content":       *chat.LMContent,
				"createdAt":     *chat.LMCreated,
				"formattedDate": formatChatShortDate(*chat.LMCreated),
				"isFromMe":      *chat.LMSender == userID,
				"isRead":        *chat.LMRead,
			}
		} else {
			lastMsg = nil
		}
		la := chat.LastMessageAt
		if la.IsZero() {
			la = chat.CreatedAt
		}
		out = append(out, map[string]any{
			"id":               chat.ID,
			"isModerationChat": chat.IsModerationChat,
			"product":          product,
			"companion": map[string]any{
				"id": compID, "fullName": compName, "phoneNumber": compPhone, "avatar": nil,
			},
			"lastMessage":  lastMsg,
			"unreadCount":  unread,
			"lastActivity": formatChatShortDate(la),
			"createdAt":    chat.CreatedAt,
		})
	}
	return out, nil
}

func (s *ChatService) GetChatInfo(ctx context.Context, chatID, userID int32) (map[string]any, error) {
	c, err := s.repo.GetChatFull(ctx, chatID)
	if errors.Is(err, repository.ErrNotFound) {
		return nil, &AppError{404, "Чат не найден"}
	}
	if err != nil {
		return nil, err
	}
	if c.BuyerID != userID && c.SellerID != userID {
		return nil, &AppError{403, "Нет доступа к этому чату"}
	}
	isBuyer := c.BuyerID == userID
	var comp map[string]any
	if isBuyer {
		comp = map[string]any{"id": c.SellerID, "fullName": c.SellerName, "phoneNumber": c.SellerPhone, "role": "seller"}
	} else {
		comp = map[string]any{"id": c.BuyerID, "fullName": c.BuyerName, "phoneNumber": c.BuyerPhone, "role": "buyer"}
	}
	if c.IsModerationChat {
		comp["fullName"] = "Модерация платформы"
		comp["phoneNumber"] = ""
		comp["role"] = "moderation"
	}
	var product any
	if c.JoinProductID != nil {
		var img any
		if len(c.ProductImages) > 0 {
			img = c.ProductImages[0]
		} else {
			img = nil
		}
		product = map[string]any{
			"id":          *c.JoinProductID,
			"name":        c.ProductName,
			"price":       c.ProductPrice,
			"image":       img,
			"description": c.ProductDesc,
		}
	} else {
		product = nil
	}
	ub, us, err := s.repo.ChatUnreadCounts(ctx, chatID)
	if err != nil {
		return nil, err
	}
	unreadCnt := us
	if isBuyer {
		unreadCnt = ub
	}
	return map[string]any{
		"id":               c.ID,
		"isModerationChat": c.IsModerationChat,
		"product":          product,
		"companion":        comp,
		"isUserBuyer":      isBuyer,
		"unreadCount":      unreadCnt,
		"createdAt":        c.CreatedAt,
	}, nil
}

// GetChatMessages — как Nest getChatMessages: DESC из БД, затем reverse; page/limit по умолчанию 1 и 50.
func (s *ChatService) GetChatMessages(ctx context.Context, chatID, userID int32, page, limit int) (map[string]any, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 {
		limit = 50
	}
	buyerID, sellerID, err := s.repo.ChatParticipants(ctx, chatID)
	if errors.Is(err, repository.ErrNotFound) {
		return nil, &AppError{404, "Чат не найден"}
	}
	if err != nil {
		return nil, err
	}
	if buyerID != userID && sellerID != userID {
		return nil, &AppError{403, "Нет доступа к этому чату"}
	}
	offset := (page - 1) * limit
	rows, total, err := s.repo.ListMessagesPage(ctx, chatID, limit, offset)
	if err != nil {
		return nil, err
	}
	slices.Reverse(rows)
	msgs := make([]map[string]any, 0, len(rows))
	for _, m := range rows {
		var rel any
		if m.RelProductID != nil {
			rel = map[string]any{
				"id":            *m.RelProductID,
				"name":          m.RelName,
				"images":        m.RelImages,
				"price":         m.RelPrice,
				"moderateState": m.RelModerate,
			}
		} else {
			rel = nil
		}
		msgs = append(msgs, map[string]any{
			"id":             m.ID,
			"content":        m.Content,
			"senderId":       m.SenderID,
			"sender":         map[string]any{"id": m.SenderID, "fullName": m.SenderName},
			"relatedProduct": rel,
			"isFromMe":       m.SenderID == userID,
			"isRead":         m.IsRead,
			"readAt":         m.ReadAt,
			"createdAt":      m.CreatedAt,
			"timeString":     messageTimeString(m.CreatedAt),
		})
	}
	var pages int32
	if limit > 0 {
		pages = int32(math.Ceil(float64(total) / float64(limit)))
	}
	return map[string]any{
		"messages": msgs,
		"pagination": map[string]any{
			"page":  page,
			"limit": limit,
			"total": total,
			"pages": pages,
		},
	}, nil
}

// SendMessage — как Nest chat.sendMessage (для REST при необходимости и для Socket.IO).
func (s *ChatService) SendMessage(ctx context.Context, chatID, senderID int32, content string) (map[string]any, error) {
	content = strings.TrimSpace(content)
	if content == "" {
		return nil, &AppError{400, "Пустое сообщение"}
	}
	msgID, senderName, createdAt, _, _, err := s.repo.InsertChatMessage(ctx, chatID, senderID, content)
	if errors.Is(err, repository.ErrNotFound) {
		return nil, &AppError{404, "Чат не найден"}
	}
	if errors.Is(err, repository.ErrForbiddenChat) {
		return nil, &AppError{403, "Нет доступа к этому чату"}
	}
	if err != nil {
		return nil, err
	}

	s.sendAsyncChatMessageNotification(chatID, senderID, senderName, content)

	return map[string]any{
		"id":         msgID,
		"content":    content,
		"senderId":   senderID,
		"sender":     map[string]any{"id": senderID, "fullName": senderName},
		"isRead":     false,
		"createdAt":  createdAt,
		"timeString": messageTimeString(createdAt),
	}, nil
}

func (s *ChatService) sendAsyncChatMessageNotification(chatID, senderID int32, senderName, content string) {
	if s == nil || s.users == nil {
		return
	}
	go func() {
		ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
		defer cancel()

		full, err := s.repo.GetChatFull(ctx, chatID)
		if err != nil || full == nil {
			return
		}

		recipientID := full.SellerID
		if senderID == full.SellerID {
			recipientID = full.BuyerID
		}

		recipient, err := s.users.FindUserByID(ctx, recipientID)
		if err != nil || recipient == nil {
			return
		}

		email := strings.TrimSpace(strings.ToLower(recipient.Email))
		if email == "" || strings.HasSuffix(email, "@oauth.local") || !strings.Contains(email, "@") {
			return
		}

		productTitle := "объявлению"
		if full.ProductName != nil && strings.TrimSpace(*full.ProductName) != "" {
			productTitle = fmt.Sprintf("товару «%s»", strings.TrimSpace(*full.ProductName))
		}

		subject := fmt.Sprintf("Новое сообщение от %s", senderName)
		baseUrl := strings.TrimRight(s.cfg.BaseURL, "/")
		if baseUrl == "" {
			baseUrl = "http://localhost:3000"
		}
		body := fmt.Sprintf(`
			<div style="font-family: Arial, sans-serif; padding: 20px; color: #333;">
				<h2 style="color: #2563eb;">Новое сообщение в чате</h2>
				<p>Здравствуйте, <b>%s</b>!</p>
				<p>Пользователь <b>%s</b> прислал вам сообщение по %s:</p>
				<blockquote style="border-left: 4px solid #2563eb; margin: 16px 0; padding: 10px 16px; background-color: #f3f4f6; border-radius: 4px;">
					%s
				</blockquote>
				<p style="margin-top: 24px;">
					<a href="%s/profile/messages/%d" style="background-color: #2563eb; color: #ffffff; text-decoration: none; padding: 10px 18px; border-radius: 6px; display: inline-block; font-weight: bold;">
						Перейти к переписке
					</a>
				</p>
			</div>
		`, recipient.FullName, senderName, productTitle, content, baseUrl, chatID)

		_ = mail.SendHTMLSmart(
			s.cfg.SMTPHost, s.cfg.SMTPPort, s.cfg.SMTPUser, s.cfg.SMTPPassword,
			s.cfg.SMTPFrom, email, subject, body,
			s.cfg.SMTPSecure, s.cfg.SMTPTLSInsecure,
		)
	}()
}

// MarkMessagesAsRead — как Nest markMessagesAsRead.
func (s *ChatService) MarkMessagesAsRead(ctx context.Context, chatID, userID int32) error {
	err := s.repo.MarkChatMessagesRead(ctx, chatID, userID)
	if errors.Is(err, repository.ErrNotFound) {
		return &AppError{404, "Чат не найден"}
	}
	if errors.Is(err, repository.ErrForbiddenChat) {
		return &AppError{403, "Нет доступа к этому чату"}
	}
	return err
}
